### PR TITLE
Instructor: recycle bin feature: allow restore and delete sessions in instructor feedback sessions page #8906

### LINKED ISSUE #8906: Instructor: recycle bin feature: allow restore and delete sessions in instructor feedback sessions page
**Environment:**
`recycling-bin` branch

**Description of feature/enhancement:**
- Create a table for deleted feedback sessions in instructor feedback sessions page
- Display all deleted feedback sessions in this table
- Allow users to restore selected/all feedback session from Recycle Bin
- Allow users to permanently delete selected/all sessions from Recycle Bin

### PR DESCRIPTION
**Pull request by dalessr**
Fixes #8906 

**Outline of Solution:**

- Created a table in instructor feedback session page
- Added a new attribute `deletedTime` in `FeddbackSession.java` and `FeedbackSessionAttributes.java`
- Setted `deletedTime` value instead of actually deleting the session when `deleteSession` method is called
- Display all deleted feedback sessions in instructor feedback sessions page
- Added `Delete Permanently` button for each column to allow users to delete select session
- Added `Delete All Sessions` button to allow users to delete all sessions
- Added `Restore` button for each column to allow users to restore select session
- Added `Restore All Sessions` button to allow users to restore all sessions
- Disallowed students to see the session after it has been moved to Recycle Bin
- Added & modified confirmation prompt, tooltips and messages for deleted sessions table
- Modify tests relevant to soft-deletion of sessions

**To Do:**

- Disable deletion confirmation prompt in `Instructor Home Page` and `Instructor Feedback Sessions Page`
- Add check box beside each column to allow multiple sessions restore/delete
- Add tests for actions in deleted sessions table
- Change the two tables in `Instructor Feedback Sessions Page` to panels

**Current Recycle Bin UI:**

8906-image-1

8906-image-2

8906-image-3

### CODE REVIEW

**Comment by damithc** [comment]
on a related note, you may want to update the  current 'delete success' status message to mention that the course/session has been moved to the recycle bin and can be permanently deleted or restored from there.

**Comment by dalessr** [comment]
@damithc Yeah I have changed the status messages slightly. For status message to users, I wrote something like `The course ... has been deleted. You can restore it from the deleted courses table below.` Does it also have to be that specific for status message to admin?

**Comment by damithc** [comment]
> @damithc Yeah I have changed the status messages slightly. For status message to users, I wrote something like The course ... has been deleted. You can restore it from the deleted courses table below. Does it also have to be that specific for status message to admin?

Sounds good. No need to update the status message to admin.

**Comment by dalessr** [comment]
@damithc Okay sure.
@joanneong @darrenwee So far there's no need to review this PR :)

**Comment by dalessr** [comment]
@joanneong @darrenwee This PR is ready for review once #8853 is merged!

File: src/main/webapp/WEB-INF/tags/instructor/home/feedbackSessionActions.tag [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        <a href="javascript:;"
+            data-actionlink="${actions.sessionResendPublishedEmailPageLink}"
+            class="session-resend-published-email-for-test"
+            data-courseid="${actions.courseId}"
+            data-fsname="${actions.fsName}"
+            data-toggle="modal"
+            data-target="#resendPublishedEmailModal">
+          Resend link to view results
+        </a>
+      </li>
```
Comment (joanneong) [inline src/main/webapp/WEB-INF/tags/instructor/home/feedbackSessionActions.tag]: Should the `<li>` part be indented one level further than the `<c>` tag?

File: src/main/webapp/dev/js/common/instructor.js [hunk 2]
```diff
@@ -310,13 +310,14 @@ function bindDeleteButtons() {
         const courseId = $button.data('courseid');
         const feedbackSessionName = $button.data('fsname');
 
-        const messageText = `Are you sure you want to delete the feedback session ${feedbackSessionName} in ${courseId}?`;
+        const messageText = `Are you sure you want to delete the feedback session ${feedbackSessionName} in ${courseId}? `
+                + 'This action can be reverted by going to the "sessions" tab and restoring the desired session(s).';
```
Comment (joanneong) [inline src/main/webapp/dev/js/common/instructor.js]: Nit-picking as usual: the "*Sessions*" tab

File: src/test/java/teammates/test/cases/action/InstructorFeedbackDeleteActionTest.java [hunk 3]
```diff
@@ -72,6 +72,9 @@ protected void testAccessControl() throws Exception {
         verifyUnaccessibleWithoutModifySessionPrivilege(submissionParams);
         verifyAccessibleForInstructorsOfTheSameCourse(submissionParams);
 
+        //delete the previous entity
+        FeedbackSessionsLogic.inst().deleteFeedbackSessionCascade(fs.getSessionName(), fs.getCourseId());
+
```
Comment (joanneong) [inline src/test/java/teammates/test/cases/action/InstructorFeedbackDeleteActionTest.java]: Rather than deleting the previous entity, why  not just remove the next part on recreating the same entity? Since nothing about fs was actually modified.

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackSessionsPageUiTest.java [hunk 4]
```diff
@@ -586,12 +586,11 @@ private void testDeleteAction() throws Exception {
...
-        feedbackPage = getFeedbackPageForInstructor(idOfInstructorWithSessions);
-        feedbackPage.clickAndCancel(feedbackPage.getDeleteLink(courseId, sessionName));
+        feedbackPage = getFeedbackPageForInstructor(idOfInstructorWithSessions).deleteSession(courseId, sessionName);
+        assertTrue(feedbackPage.getTextsForAllStatusMessagesToUser()
+                .contains(Const.StatusMessages.FEEDBACK_SESSION_MOVED_TO_RECYCLE_BIN));
         assertNotNull("session should not have been deleted",
                       BackDoor.getFeedbackSession(courseId, sessionName));
-
-        feedbackPage.clickAndConfirm(feedbackPage.getDeleteLink(courseId, sessionName));
         feedbackPage.verifyHtmlMainContent("/instructorFeedbackDeleteSuccessful.html");
```
Comment (joanneong) [inline src/test/java/teammates/test/cases/browsertests/InstructorFeedbackSessionsPageUiTest.java]: Hmm for soft delete you can also test cancelling the soft delete as well (since it is possible to cancel)

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackEditPage.java [hunk 5]
```diff
@@ -1304,7 +1304,7 @@ public String getFeedbackSessionEndTimeValue() {
     }
 
     public InstructorFeedbackSessionsPage deleteSession() {
-        clickAndConfirm(getDeleteSessionLink());
+        click(getDeleteSessionLink());
```
Comment (joanneong) [inline src/test/java/teammates/test/pageobjects/InstructorFeedbackEditPage.java]: I think it's still reasonable to have a `clickAndConfirm` for soft-deletes.

File: src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java [hunk 6]
```diff
@@ -435,6 +472,26 @@ public void setCreatedTime(Instant createdTime) {
         this.createdTime = createdTime;
     }
 
+    public Instant getDeletedTime() {
+        return deletedTime;
+    }
+
+    public void setDeletedTime() {
```
Comment (darrenwee) [inline src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: Add a javadoc to this setter. It is not immediately clear that this sets the deleted time to now. If someone had to use this setter, they would need to click into the method to see the difference (which is annoying and not in the spirit of OOP)

File: src/main/java/teammates/ui/pagedata/InstructorFeedbackSessionsPageData.java [hunk 7]
```diff
@@ -221,6 +238,17 @@ public FeedbackSessionsCopyFromModal getCopyFromModal() {
...
+    public boolean isInstructorAllowedToModify() {
+        for (int i = 0; i < recoveryFsList.getRows().size(); i++) {
+            RecoveryFeedbackSessionsTableRow feedbackSession = recoveryFsList.getRows().get(i);
+            if (!instructorsForCourses.get(feedbackSession.getCourseId())
+                    .isAllowedForPrivilege(Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_SESSION)) {
+                return false;
+            }
+        }
+        return true;
+    }
```
[the remark this answers is not in the document]
Comment (darrenwee) [reply src/main/java/teammates/ui/pagedata/InstructorFeedbackSessionsPageData.java]: If it is used, use streams in this method 😄

File: src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java [hunk 8]
```diff
@@ -427,6 +429,41 @@ public void setInstructions(Text instructions) {
         this.instructions = instructions;
     }
 
+    public String getCreatedTimeDateString() {
+        return TimeHelper.formatDateForInstructorCoursesPage(createdTime, timeZone);
```
[the remark this answers is not in the document]
Comment (dalessr) [reply src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: Oh I just follows the format of naming in `CourseAttributes.java`. But the attribute name there is `createdAt` ->`getCreatedAtDateString()`... Yeah I agree that the name here is a bit confusing. Got any suggestions? Or shall we just leave it?

File: src/main/java/teammates/logic/api/Logic.java [hunk 9]
```diff
@@ -1265,6 +1265,24 @@ public FeedbackSessionDetailsBundle getFeedbackSessionDetails(String feedbackSes
...
+        Assumption.assertNotNull(instructor);
+        return feedbackSessionsLogic.getRecoveryFeedbackSessionsListForInstructor(instructor);
+    }
+
+    public List<FeedbackSessionAttributes> getRecoveryFeedbackSessionsListForInstructors(
+            List<InstructorAttributes> instructorList) {
+        Assumption.assertNotNull(instructorList);
+        return feedbackSessionsLogic.getRecoveryFeedbackSessionsListForInstructors(instructorList);
+    }
+
```
[the remark this answers is not in the document]
Comment (dalessr) [reply src/main/java/teammates/logic/api/Logic.java]: `getRecoveryFeedbackSessionsListForInstructor()` is used for checking instructor's privilege of deleting/restoring all sessions action in Recycle Bin. With this method, the logic of checking for each session becomes clearer and more intuitive. So I prefer to keep this method :)

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 10]
```diff
@@ -1473,6 +1511,58 @@ public void deleteFeedbackSessionCascade(String feedbackSessionName, String cour
...
+        FeedbackSessionAttributes feedbackSession = fsDb.getFeedbackSession(courseId, feedbackSessionName);
+        feedbackSession.resetDeletedTime();
+        fsDb.updateFeedbackSession(feedbackSession);
+    }
+
+    private void restoreFeedbackSessionFromRecovery(FeedbackSessionAttributes feedbackSession)
+            throws InvalidParametersException, EntityDoesNotExistException {
+        feedbackSession.resetDeletedTime();
+        fsDb.updateFeedbackSession(feedbackSession);
+    }
```
[the remark this answers is not in the document]
Comment (dalessr) [reply src/main/java/teammates/logic/core/FeedbackSessionsLogic.java]: My bad :(

File: src/main/java/teammates/ui/controller/InstructorFeedbackRestoreAllRecoverySessionsAction.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+                        Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_SESSION);
+            }
+        }
+
+        try {
+            /* Restore all sessions and setup status to be shown to user and admin */
+            logic.restoreAllFeedbackSessionsFromRecovery(instructorList);
+            String statusMessage = Const.StatusMessages.FEEDBACK_SESSION_ALL_RESTORED;
+            statusToUser.add(new StatusMessage(statusMessage, StatusMessageColor.SUCCESS));
+            statusToAdmin = "All sessions restored";
```
[the remark this answers is not in the document]
Comment (dalessr) [reply src/main/java/teammates/ui/controller/InstructorFeedbackRestoreAllRecoverySessionsAction.java]: Not really...admin can actually see who is doing the action:

8907-image-21

And that's also the reason why `statusToAdmin` in all the other actions never include instructor's information :)

File: src/main/java/teammates/ui/pagedata/InstructorFeedbackSessionsPageData.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (dalessr) [reply src/main/java/teammates/ui/pagedata/InstructorFeedbackSessionsPageData.java]: @joanneong Yeah it's used in `recoveryFeedbackSessionsPanel.tag` to check whether `Delete All/Restore All` buttons should be disabled.
@darrenwee will do.

File: src/main/webapp/WEB-INF/tags/instructor/home/feedbackSessionActions.tag (hunk 1 above)
Comment (dalessr) [reply src/main/webapp/WEB-INF/tags/instructor/home/feedbackSessionActions.tag]: Oh didn't notice that. I just copy paste `feedbacks/feedbackSessionActions.tag`. Will change both.

File: src/main/webapp/dev/js/common/instructor.js (hunk 2 above)
Comment (dalessr) [reply src/main/webapp/dev/js/common/instructor.js]: Will also change the "courses" below :)

File: src/test/java/teammates/test/cases/browsertests/InstructorFeedbackSessionsPageUiTest.java (hunk 4 above)
Comment (dalessr) [reply src/test/java/teammates/test/cases/browsertests/InstructorFeedbackSessionsPageUiTest.java]: Oh actually in `InstructorFeedbackSessionsPage` there won't be a JavaScript confirmation for soft-deleting a session. The confirmation comes out when deleting a session from `InstructorHomePage`. This follows the behaviors of archiving courses & soft-deleting courses.

File: src/test/java/teammates/test/pageobjects/InstructorFeedbackEditPage.java (hunk 5 above)
Comment (dalessr) [reply src/test/java/teammates/test/pageobjects/InstructorFeedbackEditPage.java]: Explained as above :)

**Comment by dalessr** [comment]
@darrenwee @joanneong Changes made accordingly. Ready for review :)

File: src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (joanneong) [reply src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: @darrenwee what do you think?

[the remark this answers is not in the document]
Comment (joanneong) [reply src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: @darrenwee what do you think?

[the remark this answers is not in the document]
Comment (darrenwee) [reply src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: How about `getCreatedDateStamp` since you're using `formatDate...`?

[the remark this answers is not in the document]
Comment (dalessr) [reply src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: Then how should we name the method below it (`getCreatedTimeDateStamp()`)?

My suggestions for these three method names:
`getCreatedDateString()`
`getCreatedDateTimeStamp()`
`getCreatedDateTimeFullString()`

@darrenwee @joanneong What do yall think?

[the remark this answers is not in the document]
Comment (joanneong) [reply src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: Oh oops I relooked what I wrote, and actually my concern was with `formatDateForInstructorCoursesPage` because now you're formatting the date for instructor sessions page...sorry for the confusion :(

[the remark this answers is not in the document]
Comment (dalessr) [reply src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: Oops sorry...misunderstanding :(

[the remark this answers is not in the document]
Comment (dalessr) [reply src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java]: I think I should add a new method in `TimeHelper.java`

**Comment by dalessr** [comment]
@darrenwee Changes made due to Prof's comments in #8965 . Ready for review :)

File: src/main/java/teammates/common/util/TimeHelper.java [hunk 12]
```diff
@@ -384,6 +384,18 @@ public static String formatDateForInstructorCoursesPage(Instant instant, ZoneId
...
 
+    /**
+     * Formats a date stamp from an {@code instant} for the instructor's feedback sessions page.
+     * Example: 5 May 2017
+     *
+     * @param instant the instant to be formatted
+     * @param zoneId  the time zone to calculate local date
+     * @return the formatted date stamp string
+     */
+    public static String formatDateForInstructorSessionsPage(Instant instant, ZoneId zoneId) {
```
Comment (darrenwee) [inline src/main/java/teammates/common/util/TimeHelper.java]: Can you add a `TODO` here to rename `formatDateForInstructorCoursesPage` so that we don't have duplicated code?

You can make another PR to do the refactoring (probably change the name from `formatDateForInstructorCoursesPage` to `formatDateForInstructorPages`) so that we can remove this duplicated method.

Comment (darrenwee) [reply src/main/java/teammates/common/util/TimeHelper.java]: We don't have to do the renaming here as it would touch a lot of files in this PR that are not related to the recycling bin feature.

Comment (dalessr) [reply src/main/java/teammates/common/util/TimeHelper.java]: Sure

**Comment by dalessr** [comment]
Actually this PR has already been reviewed by Prof @damithc in #8965 . It is ready to merge into `upstream/recycling-bin` branch after #9005 has been merged & `upstream/recycling-bin` has been updated accordingly.

**Comment by dalessr** [comment]
@darrenwee @joanneong Ready to merge (so sorry I have to `git merge` again since this PR contains a lot of similar changes compared with https://github.com/TEAMMATES/teammates/commit/0c0419b5c092bfbf3c5e4a8a1f29535d594e7d3e, thus it's a disaster to resolve conflicts for almost each of the 195 commits T_T But I've rebased #8965 & #8969 onto this PR successfully)

**Comment by damithc** [comment]
This may be useful when rebasing a long branch (haven't used it myself though)
https://git-scm.com/book/en/v2/Git-Tools-Rerere

**Comment by dalessr** [comment]
It seems quite convenient. Will try that!
